x/msg_authorization/internal/keeper/keeper.go

```diff
@@ -0,0 +1,41 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/msg_authorization/internal/types"
+	"time"
+)
+
+type Keeper struct {
+	storeKey sdk.StoreKey
+	cdc      *codec.Codec
+	router   sdk.Router
+}
+
+// NewKeeper constructs a message authorisation Keeper
+func NewKeeper(storeKey sdk.StoreKey, cdc *codec.Codec, router sdk.Router) Keeper {
+	return Keeper{
+		storeKey: storeKey,
+		cdc:      cdc,
+		router:   router,
+	}
+}
+
+func (k Keeper) DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []sdk.Msg) sdk.Result {
+	//TODO
+	return sdk.Result{}
+}
+
+func (k Keeper) Grant(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, capability types.Capability, expiration time.Time) {
+	//TODO
+}
+
+func (k Keeper) Revoke(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, msgType sdk.Msg) {
+	//TODO
+}
+
+func (k Keeper) GetCapability(ctx sdk.Context, grantee sdk.AccAddress, granter sdk.AccAddress, msgType sdk.Msg) (cap types.Capability, expiration time.Time) {
```

Comment:
Why does this return the expiration? Doesn't that detail belong in the keeper - eg. returning `nil` if expired?

Reply:
Note: one issue I had in the fee delegation is that when it detects it is expired, it can try to delete the capability. However, the tx will return an error (as this msg is unauthorized) and the sdk will revert state changes, which also includes deleting the expired capability.

Hmmm....

Reply:
@aaronc how would you like me to go ahead with this

Reply:
I'm not sure what is best to be honest.

The pro to this approach is that it the expiration will usually be stored next to the capability.

An alternative is to return null if expired and have a separate method `GetCapabilityExpiration`.

Yes there is the issue that it is hard for expired capabilities to auto-delete when expired as the tx will fail.

I'm not sure if I have a better solution and am open to suggestions.

Reply:
Regarding deleting expired grants, I think for now we just ignore this, but in the future there would likely be an end blocker that does this clean-up periodically.

Reply:
Endblocker cleanup sounds good.

Unable to delete on failure is actually an architecture design of the sdk, which I generally agree with. Just this one point where I see an issue. I guess it could work in an antehandler. But no need to add that complexity here

Reply:
Okay, I see, we pass these two objects to a higher-level function that can modify the capabiltity and save it again (with same expiration)